### PR TITLE
Realize true one-shot semantics for epoll with EPOLLONESHOT

### PR DESCRIPTION
**Pull request by lihuiba**
So that all invocations to wait_for_fd_xxx() will benefit from it, especially the one in cascading engine.

### CODE REVIEW

File: io/epoll.cpp [hunk 1]
```diff
@@ -117,54 +117,57 @@ class EventEngineEPoll : public MasterEventEngine, public CascadingEventEngine,
...
             _inflight_events.resize(e.fd * 2);
         auto& entry = _inflight_events[e.fd];
-        if (e.interests & entry.interests) {
-            if (((e.interests & entry.interests & EVENT_READ) &&
-                 (entry.reader_data != e.data)) ||
-                ((e.interests & entry.interests & EVENT_WRITE) &&
-                 (entry.writer_data != e.data)) ||
-                ((e.interests & entry.interests & EVENT_ERROR) &&
-                 (entry.error_data != e.data))) {
+        auto eint = e.interests & entry.interests;
```
Comment (beef9999) [inline io/epoll.cpp:120]:
```suggestion
        auto intersection = e.interests & entry.interests;
```
变量跟下面的rm用相同的？

File: io/fd-events.h [hunk 2]
```diff
@@ -44,7 +45,7 @@ class MasterEventEngine {
      * @return 0 for success, which means event arrived in time
```
Comment (beef9999) [inline io/fd-events.h:45]: 现在不允许组合 flag了吗？那上面的注释是不是应该改一下

File: io/epoll.cpp [hunk 3]
```diff
@@ -273,35 +265,39 @@ class EventEngineEPoll : public MasterEventEngine, public CascadingEventEngine,
...
             return 0;  // Event arrived
-        } else if (ret == 0) {
-            rm_interest(event);  // Timeout
-            errno = ETIMEDOUT;
-            return -1;
-        } else {
-            rm_interest(event);  // Interrupted by other thread
-            errno = err.no;
-            return -1;
         }
```
Comment (beef9999) [inline io/epoll.cpp:281]: I think the previous if...else is more self-evident.

People would clearly see that 0 is timeout, -1 is interrupted, etc.  More educational.

File: io/epoll.cpp [hunk 4]
```diff
@@ -117,54 +117,57 @@ class EventEngineEPoll : public MasterEventEngine, public CascadingEventEngine,
...
+        if (unlikely(!e.interests)) return 0;
         auto& entry = _inflight_events[e.fd];
-        auto intersection = e.interests & entry.interests &
-                            (EVENT_READ | EVENT_WRITE | EVENT_ERROR);
+        auto intersection = e.interests & entry.interests & EVENT_RWE;
         if (intersection == 0) return 0;
 
-        auto x = (entry.interests ^= intersection) &
-                 (EVENT_READ | EVENT_WRITE | EVENT_ERROR);
+        entry.interests ^= intersection;
```
Comment (beef9999) [inline io/epoll.cpp:153]:
```suggestion
        entry.interests &= ~intersection;
```
Easier to understand

File: io/epoll.cpp [hunk 5]
```diff
@@ -117,54 +117,57 @@ class EventEngineEPoll : public MasterEventEngine, public CascadingEventEngine,
...
+        entry.interests |= e.interests;
+        auto x = entry.interests & EVENT_RWE;
         auto events = evmap.translate_bitwisely(x);
-        return ctl(e.fd, op, events);
+        if (e.interests & ONE_SHOT) {
+            events |= EPOLLONESHOT;
+            auto op = likely(eint & ONE_SHOT) ? EPOLL_CTL_MOD : EPOLL_CTL_ADD;
+            auto ret = ctl(e.fd, op, events);
+            return likely(ret >= 0) ? ret :
+                       ctl(e.fd, EPOLL_CTL_ADD, events);
```
Comment (beef9999) [inline io/epoll.cpp:139]: 这里是先尝试MOD，失败了再ADD。
如果某个旧的fd还没rm_interest，这时候有个新的add_interest带着相同的fd和ONE_SHOT来了，这能挡得住吗？

File: io/epoll.cpp (hunk 1 above)
Comment (lihuiba) [reply io/epoll.cpp]: OK

File: io/epoll.cpp (hunk 5 above)
Comment (lihuiba) [reply io/epoll.cpp]: 又加了判断

File: io/epoll.cpp (hunk 4 above)
Comment (lihuiba) [reply io/epoll.cpp]: added comments

File: io/fd-events.h (hunk 2 above)
Comment (lihuiba) [reply io/fd-events.h:45]: master engine 一直都没有组合flags，用法都是wait_for_xxx()。Cascading还是可以的。

Comment (beef9999) [reply io/fd-events.h:45]: 上面的注释： 
@param interests bitwisely OR-ed EVENT_READ, EVENT_WRITE

**Review by beef9999 — APPROVED** [review]

**Comment by beef9999** [comment]
Use the multi-connection benchmark program to run the performance test.
```
./build/output/multi-conn-perf -buf_size 512 -cascading_engine 0 -client_thread_num 8 -server_thread_num 8
./build/output/multi-conn-perf -buf_size 512 -cascading_engine 1 -client_thread_num 8 -server_thread_num 8
```

| Engine | Applied this patch | QPS |
|--------|--------|--------|
| Master | No | 1.72M |
| Cascading | No | 1.46M | 
| Master | Yes | 1.36M |
| Cascading | Yes | 1.33M |

**Comment by beef9999** [comment]
```bash
./build/output/multi-conn-perf -client_thread_num M -server_thread_num N [--cascading_engine]
```
25Gbps NIC, Kernel 5.15

|   Engine   |    Param   |  Applied patch |   QPS  |  Client CPU |  Server CPU |
|:----------:|:----------:|:--------------:|:------:|:-----------:|:-----------:|
|   Master   |  M=1, N=1  |     False      |  1300K |     100%    |      85%    |
| Cascading  |  M=1, N=1  |     False      |  380K  |     30%     |     100%    |
|   Master   |  M=1, N=1  |      True      |  1300K |     100%    |     73%     |
| Cascading  |  M=1, N=1  |      True      |  530K  |     40%     |    100%     |
|   Master   | M=4, N=16  |     False      |  930K  |    400%     |     650%    |
| Cascading  | M=4, N=16  |     False      | 1000K  |    400%     |    1570%    |
|   Master   | M=4, N=16  |      True      | 1000K  |    400%     |    610%     |
| Cascading  | M=4, N=16  |      True      | 1020K  |    400%     |    670%     |

**Pull request merged**